src/axom/primal/geometry/BezierPatch.hpp

```diff
@@ -0,0 +1,909 @@
+// Copyright (c) 2017-2023, Lawrence Livermore National Security, LLC and
+// other Axom Project Developers. See the top-level LICENSE file for details.
+//
+// SPDX-License-Identifier: (BSD-3-Clause)
+
+/*!
+ * \file BezierPatch.hpp
+ *
+ * \brief A BezierPatch primitive
+ */
+
+#ifndef AXOM_PRIMAL_BEZIERPATCH_HPP_
+#define AXOM_PRIMAL_BEZIERPATCH_HPP_
+
+#include "axom/core.hpp"
+#include "axom/slic.hpp"
+
+#include "axom/primal/geometry/NumericArray.hpp"
+#include "axom/primal/geometry/Point.hpp"
+#include "axom/primal/geometry/Vector.hpp"
+#include "axom/primal/geometry/Segment.hpp"
+#include "axom/primal/geometry/BezierCurve.hpp"
+#include "axom/primal/geometry/BoundingBox.hpp"
+#include "axom/primal/geometry/OrientedBoundingBox.hpp"
+
+#include "axom/primal/operators/squared_distance.hpp"
+
+#include <vector>
+#include <ostream>
+
+namespace axom
+{
+namespace primal
+{
+// Forward declare the templated classes and operator functions
+template <typename T>
+class BezierPatch;
+
+/*! \brief Overloaded output operator for Bezier Patches*/
+template <typename T>
+std::ostream& operator<<(std::ostream& os, const BezierPatch<T>& bPatch);
+
+/*!
+ * \class BezierPatch
+ *
+ * \brief Represents a 3D Bezier patch defined by a 2D array of control points
+ * \tparam T the coordinate type, e.g., double, float, etc.
+ * \tparam NDIMS the number of dimensions
+ *
+ * The order of a Bezier patch with (N+1)(M+1) control points is (N, M).
+ * The patch is approximated by the control points,
+ * parametrized from u=0 to u=1 and v=0 to v=1.
+ * 
+ * Contains a 2D array of positive weights to represent a rational Bezier patch.
+ * Nonrational Bezier patches are identified by an empty weights array.
+ * Algorithms for Rational Bezier curves derived from 
+ * Gerald Farin, "Algorithms for rational Bezier curves"
+ * Computer-Aided Design, Volume 15, Number 2, 1983,
+ */
+template <typename T>
+class BezierPatch
+{
+public:
+  using PointType = Point<T, 3>;
+  using VectorType = Vector<T, 3>;
+  using PlaneType = Plane<T, 3>;
+
+  using CoordsMat = axom::Array<PointType, 2>;
+  using NumsMat = axom::Array<T, 2>;
+
+  using BoundingBoxType = BoundingBox<T, 3>;
+  using OrientedBoundingBoxType = OrientedBoundingBox<T, 3>;
+  using BezierCurveType = primal::BezierCurve<T, 3>;
+
+  AXOM_STATIC_ASSERT_MSG(
+    std::is_arithmetic<T>::value,
+    "A Bezier Patch must be defined using an arithmetic type");
+
+public:
+  /*!
+   * \brief Constructor for a Bezier Patch that reserves space for
+   *  the given order of the surface
+   *
+   * \param [in] order the order of the resulting Bezier patch
```

Comment:
```suggestion
   * \param [in] {ord_u, ord_v} the order of the resulting Bezier patch
```

Do default order values indicate a nonrational Bezier patch? Regardless, the method documentation should indicate what the default values mean.

Reply:
This constructor only allocates space for a nonrational Bezier patch, the default order just means no space is reserved up front. You're right that it's unclear. I've clarified the behavior in the documentation on this function, but also in some other places that it was unclear.